[base-ui][Select] Fix Select button layout shift, add placeholder prop

mj12albert opened the pull request:
Fixes 1 of 3 in https://github.com/mui/material-ui/issues/38126#issuecomment-1665656235

This PR adds updates the default `renderValue` function to fall back to a zero-width space that prevents layout shift in SSR or when the value is empty.

Preview link (disable JS to try) https://deploy-preview-38796--material-ui.netlify.app/base-ui/react-select/#introduction

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

michaldudak commented:
We can include the placeholder and ZWS in the Select component itself, not just the demos.

mj12albert commented:
@michaldudak I've added it to the Select component and updated some of the existing demos to show the `placeholder` prop

I don't think it needs to be added to the hook, as `placeholder` would just pass through and would not be returned by a prop getter

mj12albert commented on packages/mui-base/src/Select/Select.test.tsx:
```diff
@@ -20,7 +20,7 @@ describe('<Select />', () => {
 
   const componentToTest = (
-    <Select defaultListboxOpen slotProps={{ popper: { disablePortal: true } }}>
+    <Select defaultListboxOpen defaultValue={1} slotProps={{ popper: { disablePortal: true } }}>
```
This is so that the ZWS is not rendered, or else `className="notranslate"` will cause the "test disabling class generation" part of `describeConformanceUnstyled` to fail

michaldudak reviewed (approved):
LGTM!

oliviertassinari commented:
> Preview link (disable JS to try)

Nice 👌
